doc/developer/design/20230615_webhook_source.md

```diff
@@ -0,0 +1,232 @@
+# Webhook Source
+
+### Associated
+ * [Epic] Webhook Source - [materialize#19951](https://github.com/MaterializeInc/materialize/issues/19951)
+ * Support and HTTP Post Source - [materialize#1701](https://github.com/MaterializeInc/materialize/issues/1701#issuecomment-1537908820)
+
+## Context
+
+We want to make it easier for folks to get data into Materialize. Currently the only way is via one
+of our [supported sources](https://materialize.com/docs/sql/create-source/), which depending on
+the environment, can be pretty hard to setup. A common method of publishing data from an
+application is via [webhooks](https://www.redhat.com/en/topics/automation/what-is-a-webhook). When
+some event occurrs, e.g. a new analytics event from Segment, or a new Pull Request on a GitHub
+repository, an HTTP endpoint you define will get called so you can receive the event in near
+real-time.
+
+Supporting webhooks in Materialize allows users to easily get their data into MZ, and the low
+latency of the events works very well with the premise of being an "operational" data warehouse.
+
+> **_Note_** To achieve the goal of "make it easier to get data in Materialize" we also considered
+building support for Fivetran. We may still do this in the future, but the amount of work to get
+webhooks working is significantly less, and the frequency of data is more suited to the goals of
+Materialize.
+
+## Goals
+
+1. Provide URLs that users can push data to. In Materialize this data is represented as a source
+with two columns, `headers` and `body`.
+2. Support an extensible syntax similar to `CREATE SOURCE <name> FROM HTTP REQUESTS` that allows
+users to generate the aforementioned URLs.
+3. Handle at least a total of 100 requests per-second across all `HTTP REQUESTS` sources.
+5. Code complete by August 15th, 2023.
+
+
+## Non-Goals
+
+The following are all features we want to support, but are out of scope in the immediate term.
+
+* Extracting keys from the data for `UPSERT`-like behavior.
+* Validation that the request comes from a trusted source.
+* Custom filtering of headers.
+* Manipulating the incoming data, splitting a single request into any kind of "sub-sources".
+
+## Overview
+
+Our implementation will mostly exist in the adapter, and we can re-use existing APIs on the
+`storage-client` to handle the persistence of data.
+
+We'll update the SQL parser to support our new kind of `HTTP REQUESTS` source, and add
+a new endpoint to `environmentd`'s HTTP server `/api/webhook/:id`. When a user creates this new
+kind of source, we'll generate an id, persist the new `Source` in the catalog, use the
```

Comment:
Instead of generating and returning an id, does it make sense to do `/api/webhook/<db>/<schema>/<name>`? This would have to deal with slashes and other URL-unfriendly chars in those somehow, but feels more webby?

Reply:
I had the same thought [here](https://github.com/MaterializeInc/materialize/pull/20002#discussion_r1234726035), which GitHub is weirdly hiding away by deafult.

Reply:
This sounds great to me, I updated the doc to reflect this!